hoomd/mpcd/CollisionMethod.cc

```diff
@@ -406,6 +432,158 @@ void mpcd::CollisionMethod::transferRigidBodyMomenta(uint64_t timestep)
+        ArrayHandle<double4> h_cell_vel(m_thermo->getCellVelocities(),
+                                        access_location::host,
+                                        access_mode::read);
+        ArrayHandle<double3> h_cell_energy(m_thermo->getCellEnergies(),
+                                           access_location::host,
+                                           access_mode::read);
+
+        const Scalar4 vel_mass_cell = h_cell_vel.data[bin_idx];
+        const vec3<Scalar> vel_cell(vel_mass_cell);
+        const Scalar mass_cell = vel_mass_cell.w;
+        const double3 cell_energy = h_cell_energy.data[bin_idx];
+        const unsigned int np = __double_as_int(cell_energy.z);
+
+        Scalar kinetic_energy
+            = cell_energy.x
+              + 0.5 * mass
+                    * (vel_mass.x * vel_mass.x + vel_mass.y * vel_mass.y + vel_mass.z * vel_mass.z);
+        Scalar temperature = 2 * kinetic_energy / (3 * (np - 1 + 1));
+
+        // thermostat relative velocities
+        // get the global index from the bin numbers found
+        const int3 global_cell = m_cl->getGlobalCell(bin);
+        const unsigned int global_idx = global_ci(global_cell.x, global_cell.y, global_cell.z);
+        // get the rng
+        uint16_t seed = m_sysdef->getSeed();
+        hoomd::RandomGenerator rng(
+            hoomd::Seed(hoomd::RNGIdentifier::CollisionMethod, timestep, seed),
+            hoomd::Counter(global_idx));
+
+        // the total number of degrees of freedom in the cell divided by 2
+        const double alpha = m_sysdef->getNDimensions() * (np - 1 + 1) / (double)2.;
+
+        // rescale velocity if there are other particles in the cell
+        if (alpha > 0)
+            {
+            // draw a random kinetic energy for the cell at the set temperature
+            hoomd::GammaDistribution<double> gamma_gen(alpha, 1.0);
+            const double rand_ke = gamma_gen(rng);
+
+            // generate the scale factor from the current temperature
+            // (don't use the kinetic energy of this cell, since this
+            // is total not relative to COM)
+            const double cur_ke = alpha * temperature;
+            double factor = (cur_ke > 0.) ? fast::sqrt(rand_ke / cur_ke) : 1.;
+
+            // rescale with the temperature of the cell
+            Scalar4 vel_com;
+            if (mass + mass_cell > 0)
+                {
+                vel_com.x = (vel_mass.x * mass + vel_mass_cell.x * mass_cell) / (mass + mass_cell);
+                vel_com.y = (vel_mass.y * mass + vel_mass_cell.y * mass_cell) / (mass + mass_cell);
+                vel_com.z = (vel_mass.z * mass + vel_mass_cell.z * mass_cell) / (mass + mass_cell);
+                vel_com.w = mass + mass_cell;
+
+                vel_mass.x += (vel_mass.x - vel_com.x) * factor;
+                vel_mass.y += (vel_mass.y - vel_com.y) * factor;
+                vel_mass.z += (vel_mass.z - vel_com.z) * factor;
```

Comment:
I think this should be like this, so you are scaling the velocity relative to the COM.
```suggestion
                vel_mass.x = vel_com.x + (vel_mass.x - vel_com.x) * factor;
                vel_mass.y = vel_com.y + (vel_mass.y - vel_com.y) * factor;
                vel_mass.z = vel_com.z + (vel_mass.z - vel_com.z) * factor;
```